### DISCUSSION TITLE
交互式 git 命令行工具：forgit

### DISCUSSION THREAD

**Discussion by adoyle-h** [just-talks/tech-talks#10] [3.分享创造]
https://github.com/wfxr/forgit

git + fzf，挺好用的。懂 shell 脚本的可以 fork 后自己魔改。

**Comment by tcstory** [comment]
目前我用过下面 3 种操作 git 的方式

1. oh my zsh 里面自带的 git 快捷方式
2. tig
3. nvim 加上 diffview.nvim


因为我习惯使用终端配合上 tmux, 所以, 我大概的工作流是这样的:

1. 在终端里, 进入项目中, 通过 oh my zsh 的 git 来查看分支和提交代码
2. 如果需要查看一些文件改动, 那么就打开 nvim, 配合上  diffview.nvim 来查看文件差异


tig 我很少使用

虽然我使用 webstorm 来写代码, 但是有时候查看文件 diff 的时候, 还是使用 diffview.nvim, 原因是, 我有时候需要对比不同分支的文件差异, 例如, 我会使用这样的命令 `DiffviewOpen origin/master..origin/feature`  来查看两个完全不同的分支的差异, 但是在 webstorm 中, 我无法找到能完成这样操作的方法.

**Reply by adoyle-h** [reply]
只对比不同分支的文件差异的话，其实用 git diff + [diff-so-fancy](https://github.com/so-fancy/diff-so-fancy) 就行，都不用装 nvim 和插件。比如我 diff 项目文件 `git diff 03459844~3 03459844 -- _docs/license.md  | diff-so-fancy`，

<img width="1352" alt="CleanShot 2023-03-03 at 16 22 11@2x" src="https://user-images.githubusercontent.com/1998490/222669010-76678f88-333b-4d41-9c31-26b5d8a68db9.png">

如果想要分页，在加个 `less`， `git diff 03459844~3 03459844 -- _docs/license.md  | diff-so-fancy | less`

**Reply by adoyle-h** [reply]
提这个是因为 git 命令提供了非常丰富的参数，插件或者 IDE 可能只能覆盖日常使用，并不能满足某些用户的特殊需求。

**Comment by adoyle-h** [comment]
分享我魔改后的[版本](https://github.com/adoyle-h/forgit)。

## git diff

![CleanShot 2023-03-03 at 16 24 47@2x](https://user-images.githubusercontent.com/1998490/222669585-5aabb520-fd37-4576-baf3-34867d0bb903.png)

## git log

![CleanShot 2023-03-03 at 16 26 05@2x](https://user-images.githubusercontent.com/1998490/222669815-42e1dc60-3a81-4ca0-ba81-870cfc9ea0a8.png)

**Comment by kalxd** [comment]
用tig查看分支线，用命令进行git操作，足够了。
